### ISSUE TITLE
Please make a seperate resource for public_network_access_enabled in azurerm_storage_account

### ISSUE THREAD

**Issue by AzySir** [hashicorp/terraform-provider-azurerm#18788]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

As it currently stands my azure_storage_account was blocking the `azurerm_storage_container` from being created due to the `pubic_network_access_enabled = false` being set. 

Now I can set this to true then run against the azurerm_storage_container. However this means I now need to run my script twice to get a desired result which makes 0 sense and is very static.

```
resource "azurerm_storage_account" "main" {
  provider                      = azurerm.sub
  name                          = module.launchpad.standard["storage-account"]
  resource_group_name           = azurerm_resource_group.main.name
  location                      = azurerm_resource_group.main.location
  /* public_network_access_enabled = false */
  public_network_access_enabled = true /* THIS NEEDS SOME ATTENTION */
  account_tier                  = "Standard"
  account_replication_type      = "GRS"
  enable_https_traffic_only     = true
  min_tls_version               = "TLS1_2"

  network_rules {
    bypass         = ["AzureServices"]
    default_action = "Deny"
    ip_rules       = var.permitted_ips
  }

  blob_properties {
    delete_retention_policy {
      days = 365
    }
  }

  queue_properties {
    logging {
      delete                = true #   (Required) Indicates whether all delete requests should be logged. Changing this forces a new resource.
      read                  = true #   (Required) Indicates whether all read requests should be logged. Changing this forces a new resource.
      version               = 1  #   (Required) The version of storage analytics to configure. Changing this forces a new resource.
      write                 = true #   (Required) Indicates whether all write requests should be logged. Changing this forces a new resource.
      retention_policy_days = 365  # (Optional) Specifies the number of days that logs will be retained. Changing this forces a new resource.
    }
  }

  identity {
    type = "SystemAssigned"
  }

  tags = {}

  depends_on = [azurerm_subscription.launchpad_subscription]
}
```

```
resource "azurerm_storage_container" "main" {
  provider              = azurerm.sub
  name                  = "${module.launchpad.standard["storage-account"]}-container"
  storage_account_name  = azurerm_storage_account.main.name
  container_access_type = "private"
}
```

Can we please look at this being detached so people can run this in a one run attempt and set depends on - this is way too static. 

### New or Affected Resource(s)/Data Source(s)

azurerm_storage_account, azurerm_storage_container

**Comment by magodo** [comment]
@AzySir Once #14220 is merged, you are able to create the container with `public_network_access_enabled=false`.

For now, you can use the [azapi](https://registry.terraform.io/providers/azure/azapi) provider to create the container, e.g.:

```
resource "azapi_resource" "test" {
  name      = "vhds"
  parent_id = "${azurerm_storage_account.main.id}/blobServices/default"
  type      = "Microsoft.Storage/storageAccounts/blobServices/containers@2021-04-01"
  body = jsonencode({
    properties = {
      publicAccess = "None"
    }
  })
}
```

While I'm wondering in your two step solution, after you created the container and set the `public_network_access_enabled` to `false`, are you able to successfully run `terraform plan`? Since in this case, if your machine running Terraform is not in the firewall whitelist, you are not able to read the container.

**Comment by tombuildsstuff** [comment]
@AzySir can you elaborate on why setting this to `false` doesn't let you provision the Container?

**Comment by AzySir** [comment]
@tombuildsstuff I've lost the error since but it would fail the creation. Which is fine i can understand why - however maybe I don't want to white list my current address yet so I want to add configuration on top and THEN i want to disable public network access.

I'm assuming that there is a flaw in the APIs - there is a whole networking rules resource it could just be added in there also.

@magodo yes I figure that out in the end - however still the flow is very static should be able to run it freely..

**Comment by tombuildsstuff** [comment]
@AzySir if this fails with Public Network Access Disabled then the next `terraform plan` will fail (since it can't reach the storage account) - so doing so will ultimately cause more subtle/knock-on errors the next time around.

This sounds like a duplicate of #2977, where we need to switch to using the Resource Manager APIs (to not require any network rules/public access) but can't do so since the Resource Manager and Data Plane APIs aren't feature-compatible - which is currently blocked on the Service Team. As such I'm going to close this in favour of #2977, but since splitting this out would cause the more subtle errors mentioned above unfortunately we have plans to make this a separate resource - instead this'll be fixed by #2977.

Thanks!

**Comment by AzySir** [comment]
@tombuildsstuff The level and confusion of the Azure APIs is beyond comprehensible - is there a roadmap and/or a plan somewhere where I can understand what is happening? - this seems to be the issue with several services....
